src/system/SystemLayerImplSelect.cpp

```diff
@@ -255,6 +279,30 @@ CHIP_ERROR LayerImplSelect::RequestCallbackOnPendingWrite(SocketWatchToken token
     VerifyOrReturnError(watch != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
 
     watch->mPendingIO.Set(SocketEventFlags::kWrite);
+
+#if CHIP_SYSTEM_CONFIG_USE_DISPATCH
+    if (watch->mWrSource)
+    {
+        dispatch_resume(watch->mWrSource);
```

Comment:
I don't think this is right...  For example, TCP will call `RequestCallbackOnPendingWrite` for every single packet, without doing any `ClearCallbackOnPendingWrite` in between.  So we will have multiple `dispatch_resume` calls when the dispatch source is already resumed.  You're not allowed to do that, per the API documentation.

I have not carefully checked for other failure modes in TCP and UDP here, but fundamentally assuming that their calls to RequestCallback* and ClearCallback* will be ordered in a sane way is not reasonable: those calls have an API designed around a single boolean flag, not a counter.  So we need to keep track with a boolean flag, and ensure our counter is only ever 0 or 1.

Reply:
I agree, I will update the change accordingly.